zigzag-circuit.md

```diff
@@ -0,0 +1,168 @@
+
+#### Circuit
+
+##### High Level
+
+In high level, we do 4 checks:
+
+1. **ReplicaId Check**: Check the binary representation of the ReplicaId
+2. **Inclusion Proofs Checks**: Check the inclusion proofs
+3. **Encoding Checks**: Check that the data has been correctly encoding into a replica
+4. **CommRStar Check**: Check that CommRStar has been generated correctly
+
+##### Details
+
+```go
+// 1: ReplicaId Check - Check ReplicaId is equal to its bit representation
+let ReplicaIdBits : [255]Fr = Fr_to_bits(ReplicaId)
+assert(Packed(replica_id_bits) == ReplicaId)
+
+let DataRoot, ReplicaRoot Fr
+
+for l in range LAYERS {
+  
+  if l == 0 {
+    DataRoot = CommD
+  } else {
+    DataRoot = CommR[l-1]
+  }
+
+  if l == LAYERS-1 {
+    ReplicaRoot = CommRLast
+  } else {
+    ReplicaRoot = CommR[l]
+  }
+  
+  for c in range LAYERS_CHALLENGES[l] {
+    // 2: Inclusion Proofs Checks
+    // 2.1: Check inclusion proofs for data nodes are correct
+    assert(MerkleTreeVerify(DataRoot, InclusionPath[l][c], DataProof[l][c], DataValue[l][c]))
+    // 2.2: Check inclusion proofs for replica nodes are correct
+    assert(MerkleTreeVerify(ReplicaRoot, InclusionPath[l][c], ReplicaProof[l][c], ReplicaValue[l][c]))
+    // 2.3: Check inclusion proofs for parent nodes are correct
+    for p in range PARENT_NODES {
+      assert(MerkleTreeVerify(ReplicaRoot, ParentInclusionPath[l][c][p], ParentProof[l][c][p]))
+    }
+
+    // 3: Encoding checks - Check that replica nodes have been correctly encoded
+    let ParentBits [PARENT_NODES][255]Fr
+    for p in range PARENT_NODES {
+      // 3.1: Check that each ParentValue is equal to its bit representation
+      let parent = ParentValue[l][c][p]
+      ParentBits[p] = Fr_to_bits(parent)
+      assert(Packed(ParentBits[p]) == parent)
+    }
+
+    // 3.2: Check that each key has generated correctly
+    let PreImage = ReplicaIdBits
+    for parentbits in ParentBits[l][c] {
+      PreImage.Append(parentbits)
+    }
+    let key Fr = PedersenHash(PreImage)
+    assert(PedersenHash(PreImage) == key)
+    // 3.3: Check that the data has been encoded to a replica with the right key
+    assert(ReplicaValue[l][c] == DataValue[l][c] + key)
+
+    // 4: CommRStar check - Check that the CommRStar constructed correctly
+    let hash = ReplicaId
+    for l in range LAYERS-1 {
+      hash.Append(CommR[l])
+    }
+    hash.Append(CommRLast)
+
+    assert(CommRStar == PedersenHash(hash))
+    // TODO check if we need to do packing/unpacking
+  }
+}
+```
+
+
+
+#### Verification of offline porep proof
+
+- SNARK proof check: **Check** that given the SNARK proof and the public inputs, the SNARK verification outputs true
+- Parent checks: For each `node = InclusionPath[l][c]`:
+  - **Check** that all `ParentsInclusionPaths_[l][c][0..PARENT_NODES}` are the correct parent nodes of `node` in the DRG graph.
+  - **Check** that the parent nodes are in numerical order.
```

Comment:
We also need to check for padding, since we always need the same number of parents.